records_mover/records/validated_records_hints.py

```diff
@@ -10,142 +12,175 @@
-    datetimeformat: HintDateTimeFormat
-
-    @staticmethod
-    def validate(hints: RecordsHints,
-                 fail_if_cant_handle_hint: bool) -> 'ValidatedRecordsHints':
+HintT = TypeVar('HintT')
+
+
+HintName = Literal["header-row",
+                   "field-delimiter",
+                   "compression",
+                   "record-terminator",
+                   "quoting",
+                   "quotechar",
+                   "doublequote",
+                   "escape",
+                   "encoding",
+                   "dateformat",
+                   "timeonlyformat",
+                   "datetimeformattz",
+                   "datetimeformat"]
+
+
+class Hint(Generic[HintT]):
+    def __init__(self,
+                 type_: Type[HintT],
+                 hint_name: HintName,
+                 default: HintT):
+        self.default = default
+        self.type_ = type_
+        self.hint_name = hint_name
+        self.valid_values: List[HintT] = list(get_args(type_))
+
+
+class Hints:
+    # mypy gives this when we pass the HintBlahBlah aliases in as an
+    # argument here:
+    #
+    # error: The type alias to Union is invalid in runtime context
+    #
+    # Nonetheless, the validation works.
+    datetimeformattz = Hint[HintDateTimeFormatTz](HintDateTimeFormatTz,  # type: ignore
+                                                  "datetimeformattz",
+                                                  "YYYY-MM-DD HH24:MI:SSOF")
+    datetimeformat = Hint[HintDateTimeFormat](HintDateTimeFormat,  # type: ignore
+                                              "datetimeformat",
+                                              default="YYYY-MM-DD HH24:MI:SS")  # TODO: test to see if default is typesafe
+    compression = Hint[HintCompression](HintCompression,  # type: ignore
+                                        'compression',
+                                        default=None)
+    quoting = Hint[HintQuoting](HintQuoting,  # type: ignore
+                                'quoting',
+                                default='minimal')
+    escape = Hint[HintEscape](HintEscape,  # type: ignore
+                              'escape',
+                              default='\\')
+    encoding = Hint[HintEncoding](HintEncoding,  # type: ignore
+                                  'encoding',
+                                  default='UTF8')
+    dateformat = Hint[HintDateFormat](HintDateFormat,  # type: ignore
+                                      'dateformat',
+                                      default='YYYY-MM-DD')
+    timeonlyformat = Hint[HintTimeOnlyFormat](HintTimeOnlyFormat,  # type: ignore
+                                              'timeonlyformat',
+                                              default="HH24:MI:SS")
+
+
+HintA = TypeVar('HintA')
+
+HintB = TypeVar('HintB')
+
+
+class HintValidator:
+    def __init__(self,
+                 fail_if_cant_handle_hint: bool,
+                 hints: RecordsHints):
```

Comment:
Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker